spec/firebase_app_distribution_action_spec.rb

```diff
@@ -101,4 +169,120 @@
         .to raise_error(ErrorMessage::MISSING_APP_ID)
     end
   end
+
+  describe '#run' do
+    let(:params) do
+      {
+        app: ios_app_id,
+        ipa_path: 'debug.ipa'
+      }
+    end
+
+    before(:each) { allow(File).to receive(:exist?).and_call_original }
+
+    it 'raises error if it cannot find a valid binary path' do
+      allow(File).to receive(:exist?).with('debug.ipa').and_return(false)
+
+      expect do
+        action.run(params.merge(ipa_path: nil))
+      end.to raise_error("Couldn't find binary")
+    end
+
+    it 'raises error if binary does not exist' do
+      allow(File).to receive(:exist?).with('debug.ipa').and_return(false)
+
+      expect do
+        action.run(params)
+      end.to raise_error("Couldn't find binary at path debug.ipa")
+    end
+
+    it 'raises error if contact email is nil' do
+      allow(File).to receive(:exist?).with('debug.ipa').and_return(true)
+      allow_any_instance_of(Fastlane::Client::FirebaseAppDistributionApiClient).to receive(:get_app).with(ios_app_id, 'BASIC').and_return(App.new({
+        contactEmail: nil
+      }))
+
+      expect do
+        action.run(params)
+      end.to raise_error(ErrorMessage::GET_APP_NO_CONTACT_EMAIL_ERROR)
+    end
+
+    it 'raises error if contact email is blank' do
+      allow(File).to receive(:exist?).with('debug.ipa').and_return(true)
+      allow_any_instance_of(Fastlane::Client::FirebaseAppDistributionApiClient).to receive(:get_app).with(ios_app_id, 'BASIC').and_return(App.new({
+        contactEmail: ''
+      }))
+
+      expect do
+        action.run(params)
+      end.to raise_error(ErrorMessage::GET_APP_NO_CONTACT_EMAIL_ERROR)
+    end
+
+    describe 'with android app' do
+      let(:params) do
+        {
+          app: android_app_id
+        }
+      end
+
+      describe 'when uploading an AAB' do
+        let(:params) do
+          {
+            app: android_app_id,
+            android_artifact_path: 'debug.aab'
+          }
+        end
+
+        before { allow(File).to receive(:exist?).with('debug.aab').and_return(true) }
+
+        def stub_get_app(params)
+          allow_any_instance_of(Fastlane::Client::FirebaseAppDistributionApiClient).to receive(:get_app).with(android_app_id, 'FULL').and_return(App.new({
+            projectNumber: project_number,
+            appId: android_app_id,
+            contactEmail: 'user@example.com'
+            }.merge(params)))
```

Comment:
Maybe assign the app to a variable first to avoid this tricky indentation and shorten the first line a bit?

Also, you could consider breaking up that line like so:

```suggestion
          allow_any_instance_of(Fastlane::Client::FirebaseAppDistributionApiClient)
            .to receive(:get_app)
            .with(android_app_id, 'FULL')
            .and_return(App.new({
              projectNumber: project_number,
              appId: android_app_id,
              contactEmail: 'user@example.com'
            }.merge(params)))
```

Reply:
Went the breaking it up line by line route!